velox/exec/tests/TaskTest.cpp

```diff
@@ -3273,4 +3272,71 @@ DEBUG_ONLY_TEST_F(TaskTest, taskExecutionEndTime) {
       taskStats.executionEndTimeMs - taskStats.executionStartTimeMs,
       injectedDelaySecs * 1'000);
 }
+
+// Test for Task::finishIfAllSplitGroupsAreDone()
+TEST_F(TaskTest, finishIfAllSplitGroupsAreDone) {
+  auto data = makeRowVector({
+      makeFlatVector<int64_t>(100, [](auto row) { return row; }),
+  });
+
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+
+  // Create a plan with two joins and three source nodes. One mixed grouped mode
+  // join and one non mixed grouped mode join.
+  core::PlanNodeId groupedScanNodeId;
+
+  auto planFragment = PlanBuilder(planNodeIdGenerator)
+                          .tableScan(asRowType(data->type()))
+                          .capturePlanNodeId(groupedScanNodeId)
+                          .project({"c0 % 5 as k", "c0"})
+                          .singleAggregation({"k"}, {"sum(c0)", "avg(c0)"})
+                          .planFragment();
+
+  planFragment.executionStrategy = core::ExecutionStrategy::kGrouped;
+  planFragment.groupedExecutionLeafNodeIds.emplace(groupedScanNodeId);
+  planFragment.numSplitGroups = 2;
+
+  // Case 1: empty task with 'no more splits' message arriving before the start.
+  auto task = Task::create(
+      "task-grouped-execution-empty",
+      folly::copy(planFragment),
+      0,
+      core::QueryCtx::create(driverExecutor_.get()),
+      Task::ExecutionMode::kParallel);
+
+  // Emulate 'no more splits message' before the task is started.
+  task->noMoreSplits(groupedScanNodeId);
+
+  // Task would not be in the kFinished state even though all splits are done.
+  EXPECT_EQ(task->state(), TaskState::kRunning);
+
+  task->start(1);
+
+  // No more splits will be there, the task has done everything.
```

Comment:
maybe check state before the call first.

Reply:
Ah, you mean not before the start() call, but before the finishIfAllSplitGroupsAreDone()...
The code reference does not show this well.

Added the check.